### ISSUE TITLE
Improve CPU memory efficiency

### ISSUE THREAD

**Issue by yousefmoazzam** [DiamondLightSource/httomo#170]
Generally, we are unaware of HTTomo's RAM usage, but it's likely that it is quite high, and probably unnecessarily so at times.

For example, in one case there's been the possibility of a Hamilton node with 251GB of RAM having all its RAM being used by HTTomo when processing a 28GB dataset (parallel run with 20 MPI processes), which is not very efficient and could be improved.

**Comment by dkazanc** [comment]
thanks for the issue. It is pretty important that we also have all information how CPU memory is utilised. It will be interesting to know how memory affected with MPI runs compare to serial runs.

**Comment by yousefmoazzam** [comment]
Yeah I agree! Considering how, if RAM gets full and crashes httomo when running on full-sized data, it means that it doesn't matter how good GPU memory efficiency is - if httomo crashes due to filling up RAM, that's affecting GPU pipelines too!

**Comment by yousefmoazzam** [comment]
Leaving this as a note: in attempts to benchmark httomo with multiple GPUs, I saw that a 27GB dataset could be processed with 2 MPI processes / 2 GPUs, but with 4 MPI processes / 4 GPUs, the same node runs out of RAM (the node in question has 377GB RAM total, 365GB RAM available when idle).

From this, it certainly seems a good idea to see how the RAM usage scales with number of MPI processes, as well as with the input data size.

**Comment by dkazanc** [comment]
just to follow-up on this. I've tried to reconstruct 6Gb `uint16` dataset on 32GB RAM PC. 
I'd imagine that in `float32` converted data we can get maximum of 24-25GB during reslice. This is not the case, I was steadily around 32GB when doing reconstruction and data saving into intermediate file. At some point, when the third of all blocks were processed, the memory raised higher than the available 32GB (my suspicion) and my PC crashed. 
_Notably the memory wasn't freed at all after the reslice, while it should be?_

I guess, there is a question why we need significantly more CPU memory than the reslice operation required?
